File: Python/ByTier/B/2155_B_Abraham_s_Great_Escape.py
# Problem: CF 2155 B - Abraham's Great Escape
# https://codeforces.com/contest/2155/problem/B

# * B. Abraham's Great Escape
# *
# * Problem Description:
# * Given an n x n grid and a number k, construct a grid where exactly k starting cells allow Abraham to escape.
# * An arrow in each cell forces movement in that direction. If movement leads outside the grid, he escapes.
# *
# * Algorithm/Approach:
# * - The solution builds a pattern that controls how many cells lead to escape.
# * - First, handle impossible cases: if k = n² - 1, it's impossible because we need at least one cell
# *   that cannot escape (i.e., leads into a loop or another valid escape).
# * - Build a grid with specific arrow patterns:
# *   - Start by placing 'U' in all cells.
# *   - If k > 0 and there are remaining cells to mark for escape, use 'U' moves to create 
# *     paths leading out (by moving up from lower rows).
# *   - Then alternate between 'R' and 'L' or 'D' for other rows to control how many escape.
# *   - Special handling of row n-1 to ensure proper escaping and non-looping behavior.
# *
# * Time Complexity: O(n²) per test case — we iterate through the grid once to place arrows.
# * Space Complexity: O(n²) — used to store the output grid as strings.

from typing import List


def function(n: int, k: int) -> List[str]:
    res = []
    if k == n**2 - 1:
        return ["No"]
    res.append("Yes")
    lines = k // n  # Number of full rows of 'U' arrows
    extra = k % n   # Extra cells in the last row
    for i in range(lines):
        temp = "U" * n  # Fill full rows with 'U' (up)
        res.append(temp)
    if extra:
        if lines != n - 1:  # If not the last row
            # Fill the partial row with 'U' for 'extra' cells and 'D' for remaining
            temp = "U" * extra + "D" * (n - extra)
            res.append(temp)
            lines += 1
        else:
            # Last row case: use 'U' and 'R'/'L' to ensure correct escaping
            temp = "U" * extra + "R" * (n - extra - 1) + "L"
            res.append(temp)
            lines += 1
    for i in range(n - lines):  # Fill remaining rows with 'R' and 'L'
        temp = "R" * (n - 1) + "L"
        res.append(temp)
    return res


if __name__ == "__main__":
    t = int(input().strip())
    for _ in range(t):
        n, k = [int(x) for x in input().split()]
        for x in function(n, k):
            print(x)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/B/2155_B_Abraham_s_Great_Escape.cpp
// Problem: CF 2155 B - Abraham's Great Escape
// https://codeforces.com/contest/2155/problem/B

/*
B. Abraham's Great Escape
Problem Description:
Given an n x n grid and a number k, we need to assign directions (U, D, L, R) to each cell such that exactly k starting positions allow Abraham to escape the maze by following the arrows.

Approach:
- If k is 0 or equals n^2, we can easily make all cells point inward or all outward.
- For other values of k, we try to place k cells pointing outward in a way that maximizes escape points without overcounting.
- The idea is to set the first K cells to 'U' (up) so they escape immediately if they are at the top edge.
- After setting those K steps, fill rest with appropriate directions (R, D, L) to ensure correct behavior.

Time Complexity: O(N^2) per test case due to filling the grid.
Space Complexity: O(N^2) for storing the grid.

Algorithms/Techniques:
- Greedy assignment of arrow directions
- Grid traversal and conditional logic

*/

#include <iostream>

using namespace std;
#define ll long long
#define f first
#define s second
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  int T;
  cin >> T;
  while (T--) {
    int N, K, wyk = 0; // wyk keeps track of how many cells have been assigned the escape move 'U'
    cin >> N >> K;
    if (N * N < K || K == N * N - 1) { // Impossible cases: not enough space or only one non-escape cell
      cout << "NO\n";
      continue;
    }
    cout << "YES\n";
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < N; j++) {
        if (wyk < K) { // If we still have moves to assign as escape cells
          wyk++; // increment escape counter
          cout << "U"; // set escape direction
        } else {
          if (i != N - 1) // if not on last row
            cout << "D";
          else { // if on last row
            if (j == N - 1)
              cout << "L"; // corner case: last row, last column -> 'L'
            else
              cout << "R"; // otherwise -> 'R'
          }
        }
        if (j == N - 1)
          cout << "\n";
      }
    }
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/